src/systemtask/SystemTask.cpp

```diff
@@ -497,7 +500,7 @@ void SystemTask::OnTouchEvent() {
 }
 
 void SystemTask::PushMessage(System::Messages msg) {
-  if (msg == Messages::GoToSleep) {
+  if (msg == Messages::GoToSleep && !doNotGoToSleep) {
     isGoingToSleep = true;
   }
```

Comment:
I'm thinking this part shouldn't even be here. This function is for pushing messages, but it changes the state of system task too and `isGoingToSleep` is set to true in the message handling too so it's set twice.

Reply:
My research said it was don't there for some race condition fixing. I couldn't completely logic it away so I chose to leave it.

Reply:
So looked again, I think the whole push routine needs some adjustment since it is allowing tasks to fire before the event is processed, and im not sure that is the correct behavior, 
See the `portYIELD_FROM_ISR` call in that routine.

Reply:
Was that the race condition fixed in commit 20f5b0f

Reply:
Thats what the git blame says yes, I was on mobile when I typed this earlier and never got around to adding it thanks!

Reply:
The variable was added to fix some race condition, but it was added to both places at the same time, which looks like a mistake to me. It wasn't added to `PushMessage()` specifically to fix the issue.

Reply:
I disagree, it was added in PushMessage since after that the display task or similar can preempt the GoingToSleep event queuing up Display task stuff after its to late for it to short circuit. Really this whole SPI/i2c side of things needs an overhaul to not rely completely on the System task for their control In my opinion. Which is out of the scope of this PR. For now fixing a potential crash is the goal :)

Reply:
I agree with @Riksu9000, it looks like `isGoingToSleep` is set to `true` in `SystemTask::PushMessage()` when the message is `Messages::GoToSleep` and also when the message `Messages::GoToSleep` is handled, which looks redundant. 

We would need to cross-check, but we could probably remote the assignation in the message handler.

> I'm thinking this part shouldn't even be here.

I also agree with this. The state machine of SystemTask needs a bit of refactoring and rationalization. There are currently multiple boolean variable that represents the state, and it's getting tricky to understand and maintain it.
But as @geekbozu said, this is not the scope of this PR. 

If everyone agrees with this, could you close this conversation so that I can merge it?

Reply:
I'm of the thought that we really should not have the assignation in Push Message. However this is something to be fixed with a State machine cleanup. So for now lets shelf this and get this patch in and we can clean it up when we actually fix our state machine.